### PR TITLE
win_stat: Module cleanup

### PR DESCRIPTION
**Pull request by dagwieers**
##### ISSUE TYPE

 - Docs Pull Request

##### COMPONENT NAME

win_stat

##### ANSIBLE VERSION

v2.3

##### SUMMARY

Small cleanup.

### CODE REVIEW

**Comment by mattclay** [comment]
Closing and re-opening to trigger CI.

**Comment by dagwieers** [comment]
This relies on Windows aliases support in #20599 (this explains the CI errors)

**Comment by gundalow** [comment]
needs_revision

The integration tests for win_stat are now failing.

**Comment by dagwieers** [comment]
@gundalow It relies on Windows aliases support in #20599 (this explains the CI errors)

Can you rebuild, #20599 has been merged !

**Comment by dagwieers** [comment]
@mattclay @jhawkesworth @nitzmahone @trondhindenes @JimFicarra Please review and test this.
I need two more shipits :-)

**Comment by mattclay** [comment]
@dagwieers There are merge conflicts from a recently merged PR.

**Comment by dagwieers** [comment]
Rebased and fixed a newly introduced bug.

**Comment by jhawkesworth** [comment]
@dagwieers Only thing I would like to see added is a note in the changelog as switching get_md5 behaviour again could potentially break some playbooks.

I did consider whether it would be better to just make get_md5 and alias for get_checksum and allow 'md5' as one of the checksum algorithms, but I suppose since you can return both an md5 and some other type of checksum then this is probably the best way to retain some compatibility with older versions.

**Comment by dagwieers** [comment]
My initial patch was also doing an alias and returning both checksums for backward compatibikity. But I figured that after a recent change I keeo tge old behaviour as is.

I don't mind adding a warning to get people moved to get_checksum, however in that case md5 should become a standard selectable checksum_algorithm.

**Comment by jhawkesworth** [comment]
A warning would definitely be better than just a note in the changelog.

If I recall the original reason for moving away from md5 was that it is not compatible with the FIPS 140 standard required for some systems.  Not everyone has government-mandated security requirements though, so yes, would have to allow md5 as one of the available checksum_algorithm options.

**Comment by dagwieers** [comment]
@jhawkesworth As requested ;-)

Now, if the `get_md5:` parameter is used and no `checksum_algorithm:` parameter was used, we give a warning and fall back to `get_checksum:` with `checksum_algorithm: md5`. At some point in the future we may deprecate `get_md5`.

Please re-review !

**Comment by dagwieers** [comment]
Falling back to `get_checksum:` was no option because the output includes `md5` instead of `checksum`.
So I now only report the warning.

File: lib/ansible/modules/windows/win_stat.py [hunk 1]
```diff
@@ -175,7 +179,7 @@
             description: the target of the symbolic link, will return null if not a link or the link is broken
             return: success, path exists, file is a symbolic link
             type: string
-            sample: C:\\temp
+            sample: C:\temp
```
Comment (mattclay) [inline lib/ansible/modules/windows/win_stat.py:182]: CI failure due to module docs issue:

```
2017-01-31 05:22:03 lib/ansible/modules/windows/win_stat.py
2017-01-31 05:22:03 ============================================================================
2017-01-31 05:22:03 TRACE:
2017-01-31 05:22:03     mapping values are not allowed here
2017-01-31 05:22:03       in "win_stat.RETURN", line 182, column 22:
2017-01-31 05:22:03                     sample: C:	emp
2017-01-31 05:22:03                              ^
2017-01-31 05:22:03 ERROR: RETURN is not valid YAML. Line 182 column 22
```

Comment (dagwieers) [reply lib/ansible/modules/windows/win_stat.py:182]: Fixed. The problem was not that specific part, the whole string was not a raw string (which I thought I fixed in all modules).

**Comment by dagwieers** [comment]
This module relies on changes from #20903 (i.e. Deprecate() function)

**Comment by JimFicarra** [comment]
@dagwieers using latest 2.3_devel and using the PowerShell.ps1 module util, this is my task:
```
- name: Test win_stat
  win_stat:
    path: d:\junk.txt
    get_md5: true
```
I get the following output.  Notice the last piece which triggers an error in the callback

```
ok: [win2k12.mycompany.com] => {
    "changed": false,
    "deprecations": [
        [
            "The parameter 'get_md5' is being replaced with 'checksum_algorithm: md5'"
        ]
    ],
    "invocation": {
        "module_name": "win_stat"
    },
    "stat": {
        "attributes": "Archive",
        "checksum": "c858ec2d85f93ce26cd8f849806ed047ded2172a",
        "creationtime": 1485028897.2348542,
        "exists": true,
        "extension": ".txt",
        "filename": "junk.txt",
        "isarchive": true,
        "isdir": false,
        "ishidden": false,
        "islink": false,
        "isreadonly": false,
        "isshared": false,
        "lastaccesstime": 1485028897.2348542,
        "lastwritetime": 1485028897.328601,
        "md5": "4e3357524c0a9224e6312717000d96f3",
        "owner": "BUILTIN\\Administrators",
        "path": "D:\\junk.txt",
        "size": 29152
    }
}
 [WARNING]: Failure using method (v2_runner_on_ok) in callback plugin (<ansible.plugins.callback.default.CallbackModule object at 0x1e25150>):
deprecated() argument after ** must be a mapping, not list
```

I don't have any suggestions for correction - not sure what the deprecation structure needs to look like when you return it in the $results, but it looks like it's choking in the [CallbackBase class](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/callback/__init__.py) in this method

```
def _handle_warnings(self, res):
        ''' display warnings, if enabled and any exist in the result '''
        if C.COMMAND_WARNINGS:
            if 'warnings' in res and res['warnings']:
                for warning in res['warnings']:
                    self._display.warning(warning)
            if 'deprecations' in res and res['deprecations']:
                for warning in res['deprecations']:
                    self._display.deprecated(**warning)
```

**Comment by dagwieers** [comment]
@JimFicarra Is it possible you are reporting to the wrong PR. Warn() and Deprecate() were added in #20903 only. And the reason why it failed is because in the meantime the deprecations-list has been turned into a dictionary, sigh, see #20884

**Comment by JimFicarra** [comment]
@dagwieers - seems like I might be.  At this point I'm not sure which combination of module/module_utils should work together. If you can provide insight that would be helpful.

**Comment by JimFicarra** [comment]
Sorry for the comment spam, but it seems like there's an issue with the `islink` attribute that precedes this PR with respect to symbolic or hard links on files.  The section of code looks for `ReparsePoint` in the FileInfo object which doesn't exist for a symbolic or hard link.

```
$info = Get-Item -Force $path

$attributes = @()
    foreach ($attribute in ($info.Attributes -split ",")) {
        $attributes += $attribute.Trim()
    }

if ($attributes -contains "ReparsePoint") {
        # TODO: Find a way to differenciate between soft and junction links
        $result.stat.islink = $true
        $result.stat.isdir = $true
        # Try and get the symlink source, can result in failure if link is broken
        try {
            $result.stat.lnk_source = [Ansible.Command.SymLinkHelper]::GetSymbolicLinkTarget($path)
        } catch {
            $result.stat.lnk_source = $null
        }
```
The System.IO.FileInfo object's attributes for a hard link has just `Archive`
```
PS D:\win_acl_test> $info = Get-Item -Force D:\win_acl_test\junk.txt
PS D:\win_acl_test> $info.Attributes
Archive
```

The `D:\win_acl_test\junk.txt` file is definitely a hardlink:
```
PS D:\win_acl_test> fsutil hardlink list d:\win_acl_test\junk.txt
\junk.txt
\win_acl_test\junk.txt
```

The same happens if I point it to a symbolic link - since the object's attributes does not contain `ReparsePoint` the code that sets islink to true and tries to subsequently find the link source doesn't execute.  I'm testing this against 2012.  I have not yet run it against 2008.

Can someone else confirm this or correct me if I'm missing something?  I did not see this issue as an open or closed issue for win_stat.

**Comment by dagwieers** [comment]
@JimFicarra Please open a new issue ticket for new things. What you are reporting has nothing to do with the cleanup and thus does not belong here.

**Comment by JimFicarra** [comment]
@dagwieers I deleted the details of the islink issue and wrote up a new issue for it.

**Comment by dagwieers** [comment]
@JimFicarra Just a heads up, the Warn() and Deprecate() functions have been renamed in true Powershell style to resp. Add-Warning() and Add-DeprecationWarning(). If you're mixing stuff, expect breakage ;-)

**Comment by jhawkesworth** [comment]
Since #20903 has now been merged and the existing win file tests are passing in the ci, I think this should be included.

At some point it should probably be tweaked to use Get-FileChecksum (in powershell.ps1 instead of its own more primitive version Get-Hash).  There might be a historic reason for this that I don't now recall, but I don't see this as a reason not to merge this change.

shipit

**Comment by jhawkesworth** [comment]
bot_status

**Comment by dagwieers** [comment]
bot_broken

**Comment by gundalow** [comment]
https://github.com/dagwieers/ansible/commit/4a12f45b943eb34967f696f925336094ef7cbfa0 us the offending commit with the `@` mention.

**Comment by mattclay** [comment]
The bot appears to be working correctly now. I've removed the comment triggering broken, as well as the label.

**Comment by dagwieers** [comment]
@gundalow I was told only the title was a problem. Does anyone have a link to the actual problem ? I never heard of this issue and I can't seem to find it via Google either.

**Comment by mattclay** [comment]
@dagwieers Here's one that @nitzmahone gets notifications about: https://github.com/ansible/ansible/commit/b369ea570a6ffd05bceb3758743d28a5a31d7045

**Comment by mattclay** [comment]
Well, the most recent one anyways.

**Comment by dagwieers** [comment]
@mattclay I meant a link describing the problem. Because I don't understand why you'd be getting repeated notifications of older commits. Seems more like a bug in the system.

**Comment by sivel** [comment]
It's generally not from this repo. If someone is doing something with their own fork, and they do a merge instead of a rebase in their own, it can cause notifications.

I get these frequently from where people have referenced me in commits.

**Comment by mattclay** [comment]
@dagwieers Depending on who you ask, it's either a bug or a feature of GitHub. Those who are on the receiving end of the notifications tend to think of it as a bug. Either way, it's beyond our control, except to prohibit commits from having an `@` mention in them. The name can stay, just drop the `@`.

**Comment by dagwieers** [comment]
It's already fixed.

**Comment by dagwieers** [comment]
ready_for_review

**Review by nitzmahone — APPROVED** [review]

**Pull request merged**